### ISSUE TITLE
Proposed high level API - [merged]

### ISSUE THREAD

**Issue by jaraco** [python/importlib_metadata#133]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 19:34

_Merges api -> master_

I propose the following high level API so that users won't have to directly
import and instantiate Distribution objects.  We can certainly flesh out the
API more as we go along.

This also prepares for a Python 2 port by moving the main implementation into
a _py3.py submodule.  The port will happen later.

I've also fixed several things about the general project setup, such as
flake8, mypy, etc.  I'll improve coverage to 100% in a follow up.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 20:10

added 1 commit

<ul><li>8878274a - Only test Python 3.6-3.7 for now</li></ul>

[Compare with previous version](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15004805&start_sha=bdbe5d56261a366426418eeb325829cad27b0806)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 21:55

added 2 commits

<ul><li>6136f21a - Dang, missed some</li><li>dd96b2ad - Force Python 3.6 for flake8&#39;ing</li></ul>

[Compare with previous version](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15006596&start_sha=8878274aa5e429e9c0c0b01c08bd98a981903ba9)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 22:06

added 1 commit

<ul><li>72641e33 - Don&#39;t run codecov; we don&#39;t have a token</li></ul>

[Compare with previous version](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15006737&start_sha=dd96b2add8e23ac04780d01c109b19972bc17e41)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 22:18

added 1 commit

<ul><li>ee2d6329 - Re-establish codecov</li></ul>

[Compare with previous version](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15006857&start_sha=72641e33f159029b04d17fa528f82bbf18702b55)

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 22:50

Commented on [importlib_metadata/_py3.py line 36](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR36)



Just raise the class.

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 22:51

Commented on [importlib_metadata/_py3.py line 46](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR46)



Capitalize and end in a period.

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 22:51

Commented on [importlib_metadata/_py3.py line 45](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR45)



A comment about what each glob is supposed to match would be nice, e.g. is one for dis-info and the other for eggs?

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 22:52

Commented on [importlib_metadata/_py3.py line 53](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR53)



"month"?

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 22:55

Commented on [importlib_metadata/__init__.py line 14](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-7df830ab27be8604ebc39a6f265b2eabR14)



Will there ever be a `set_distribution()`? If not then is it worth prefacing the functions with `get_`?

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:12

Commented on [importlib_metadata/_py3.py line 36](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR36)



Yep.  I'd actually fixed this in my unpushed porting branch.  I'll fix it for this one instead.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:13

Commented on [importlib_metadata/_py3.py line 46](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR46)



Yep!

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:15

Commented on [importlib_metadata/_py3.py line 45](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR45)



Yep

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:15

Commented on [importlib_metadata/_py3.py line 53](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR53)



I think @jaraco probably meant "module".

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:17

Commented on [importlib_metadata/__init__.py line 14](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-7df830ab27be8604ebc39a6f265b2eabR14)



Good question.  I don't think there will ever be `set_*()` functions.  I'm happy to drop the `get_` prefixes.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:17

resolved all discussions

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Thanks @brettcannon - I've pushed an update.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Commented on [importlib_metadata/_py3.py line 36](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR36)



changed this line in [version 7 of the diff](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#8205d434324de6fae88e178240a23eca3b6a4e1f_36_36)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Commented on [importlib_metadata/_py3.py line 46](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR46)



changed this line in [version 7 of the diff](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#8205d434324de6fae88e178240a23eca3b6a4e1f_46_45)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Commented on [importlib_metadata/_py3.py line 45](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR45)



changed this line in [version 7 of the diff](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#8205d434324de6fae88e178240a23eca3b6a4e1f_45_45)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Commented on [importlib_metadata/_py3.py line 53](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-5e745973e74d9e235bc0ad7eb6b8987dR53)



changed this line in [version 7 of the diff](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#8205d434324de6fae88e178240a23eca3b6a4e1f_53_55)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

Commented on [importlib_metadata/__init__.py line 14](https://github.com/python/importlib_metadata/compare/6dcfe7fbb00935b3e735cfa6ae543e5fac54109d..ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#diff-7df830ab27be8604ebc39a6f265b2eabR14)



changed this line in [version 7 of the diff](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4#b2939a55ff866ea1adb6e9d0528b2fdc704813e9_14_14)

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 25, 2018, 23:18

added 1 commit

<ul><li>533a1edc - Respond to review comments</li></ul>

[Compare with previous version](https://gitlab.com/python-devs/importlib_metadata/merge_requests/1/diffs?diff_id=15007561&start_sha=ee2d63294084a65ad8eb1e6e0f2618f765d6c2f4)

**Comment by jaraco** [comment]
In GitLab by [@brettcannon](https://gitlab.com/brettcannon) on May 25, 2018, 23:38

+1 from me, but I didn't click the "Approve" button since I will leave that to @jaraco .

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on May 30, 2018, 01:15

What do you think @jaraco ?

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:04

For some reason, I haven't been getting notifications on GitLab; otherwise I would have responded to this more promptly. I've updated my notification e-mail and hope that not to be the case going forward. Now - for the review...

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:10

approved this merge request

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:11

merged

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:11

mentioned in commit 5f4019a0b00c91e5e2911614720cadacb8a7472f

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:35

mentioned in commit 441d03c32e4d6b673566217047cf2ce5fe7ead9a

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jul 15, 2018, 01:47

I realize now I failed to catch that we hadn't updated the docs (README) for this change, which I did in 441d03c. Making that change, I realized there's some inconsistency in the API - one can construct a Distribution from a module or distribution name, but it's not possible to construct a Distribution from a module name, which a very common ([and advertised](https://gitlab.com/python-devs/importlib_metadata/blob/441d03c32e4d6b673566217047cf2ce5fe7ead9a/README.rst#L42)) pattern demands.

Should the high-level API also support this input? Or should we leave the docs as-is and expect projects to use the low-level API to get a version for a module by name?
